CFO: Balanced task scheduling

sentientwaffle opened the pull request:
## Bug

Right now the CFO chooses which fuzzer to run via a probability-proportional sampling from the task weights. The intent is that we can prioritize certain fuzzers (e.g. the VOPR).

But the sampling doesn't take into account how long fuzzers tend to run. This leads to long-running fuzzers spending more than their fair share of time on the CPU.

More specifically, CFO spends the vast majority of its time on LSM fuzzers, starving the VOPR:

- In my local tests, when fuzzing main + 2 PR's, the CFO spends 50-95% of its time fuzzing main's LSM. (Lots of variance).
- Only 1-10% of its time was spent on main's VOPRs.

(Less quantitatively, if you `ssh` into a CFO box and run `top`, it is rare to see the VOPR running.)

## Fix

Change how we sample tasks to fuzz:

- Track estimate cumulative runtime for each task.
- When selecting a task to fuzz, choose the task with the minimum `task_cumulative_runtime / task_weight`.
- When a new task is added to the pool (e.g. because we discover a new PR to fuzz), initialize its runtime to the minimum runtime of any task.
  - This ensures that a new task doesn't starve the old tasks by starting with zero runtime.
  - It also allows the old tasks to reuse their old runtimes. Tallying runtimes over larger timespans gives us more accurate estimates and thus more balanced scheduling.

The above algorithm is inspired by Linux's "Completely Fair Scheduler".

Note that the runtime estimate doesn't exactly correlate to "real" time:
- we increment the runtime when starting a fuzzer,
- the runtime doesn't count time that the fuzzer is running while the CFO is also compiling another task, and
- if a fuzzer completes in less than 1 second (the CFO's loop-sleep) then it still counts as 1 second of runtime.

## Follow-up

We should probably adjust the weights as well, e.g. not run the canary so often.

## Measurements

These are just 20 minute samples (with concurrency=1).

<details>
<summary>Cumulative estimated runtimes, pre-fair-scheduling</summary>

```
debug: commit=bb40482 fuzzer=canary                   runtime=11
debug: commit=bb40482 fuzzer=ewah                     runtime=5
debug: commit=bb40482 fuzzer=lsm_cache_map            runtime=206
debug: commit=bb40482 fuzzer=lsm_forest               runtime=302
debug: commit=bb40482 fuzzer=lsm_manifest_level       runtime=15
debug: commit=bb40482 fuzzer=lsm_manifest_log         runtime=15
debug: commit=bb40482 fuzzer=lsm_scan                 runtime=9
debug: commit=bb40482 fuzzer=lsm_segmented_array      runtime=13
debug: commit=bb40482 fuzzer=lsm_tree                 runtime=189
debug: commit=bb40482 fuzzer=storage                  runtime=13
debug: commit=bb40482 fuzzer=vopr_lite                runtime=27
debug: commit=bb40482 fuzzer=vopr_testing_lite        runtime=27
debug: commit=bb40482 fuzzer=vopr_testing             runtime=45
debug: commit=bb40482 fuzzer=vopr                     runtime=49
debug: commit=bb40482 fuzzer=vsr_free_set             runtime=12
debug: commit=bb40482 fuzzer=vsr_journal_format       runtime=5
debug: commit=bb40482 fuzzer=vsr_superblock_quorums   runtime=17
debug: commit=bb40482 fuzzer=vsr_superblock           runtime=13
debug: commit=3f4fe4f fuzzer=canary                   runtime=37
debug: commit=3f4fe4f fuzzer=vopr_lite                runtime=58
debug: commit=d7167fd fuzzer=canary                   runtime=39
debug: commit=d7167fd fuzzer=vopr_lite                runtime=88
debug: commit=d7167fd fuzzer=vopr                     runtime=278
```

</details>

<details>
<summary>Cumulative estimated runtimes, post-fair-scheduling</summary>


(Note that `lsm_tree` only ran once, just for a long time:)

```
debug: commit=bb40482 fuzzer=canary                   runtime=35
debug: commit=bb40482 fuzzer=ewah                     runtime=35
debug: commit=bb40482 fuzzer=lsm_cache_map            runtime=34
debug: commit=bb40482 fuzzer=lsm_forest               runtime=64
debug: commit=bb40482 fuzzer=lsm_manifest_level       runtime=35
debug: commit=bb40482 fuzzer=lsm_manifest_log         runtime=34
debug: commit=bb40482 fuzzer=lsm_scan                 runtime=41
debug: commit=bb40482 fuzzer=lsm_segmented_array      runtime=34
debug: commit=bb40482 fuzzer=lsm_tree                 runtime=104
debug: commit=bb40482 fuzzer=storage                  runtime=34
debug: commit=bb40482 fuzzer=vopr_lite                runtime=68
debug: commit=bb40482 fuzzer=vopr_testing_lite        runtime=69
debug: commit=bb40482 fuzzer=vopr_testing             runtime=84
debug: commit=bb40482 fuzzer=vopr                     runtime=68
debug: commit=bb40482 fuzzer=vsr_free_set             runtime=34
debug: commit=bb40482 fuzzer=vsr_journal_format       runtime=35
debug: commit=bb40482 fuzzer=vsr_superblock_quorums   runtime=35
debug: commit=bb40482 fuzzer=vsr_superblock           runtime=35
debug: commit=3f4fe4f fuzzer=canary                   runtime=121
debug: commit=3f4fe4f fuzzer=vopr_lite                runtime=239
debug: commit=d7167fd fuzzer=canary                   runtime=119
debug: commit=d7167fd fuzzer=vopr_lite                runtime=238
debug: commit=d7167fd fuzzer=vopr                     runtime=238
```

sentientwaffle committed:
CFO: Balanced task scheduling

# Bug

Right now the CFO chooses which fuzzer to run via a probability-proportional sampling from the task weights. The intent is that we can prioritize certain fuzzers (e.g. the VOPR).

But the sampling doesn't take into account how long fuzzers tend to run. This leads to long-running fuzzers spending more than their fair share of time on the CPU.

More specifically, CFO spends the vast majority of its time on LSM fuzzers, starving the VOPR:

- In my local tests, when fuzzing main + 2 PR's, the CFO spends 50-95% of its time fuzzing main's LSM. (Lots of variance).
- Only 1-10% of its time was spent on main's VOPRs.

(Less quantitatively, if you `ssh` into a CFO box and run `top`, it is rare to see the VOPR running.)

# Fix

Change how we sample tasks to fuzz:

- Track estimate cumulative runtime for each task.
- When selecting a task to fuzz, choose the task with the minimum `task_cumulative_runtime / task_weight`.
- When a new task is added to the pool (e.g. because we discover a new PR to fuzz), initialize its runtime to the minimum runtime of any task.
  - This ensures that a new task doesn't starve the old tasks by starting with zero runtime.
  - It also allows the old tasks to reuse their old runtimes. Tallying runtimes over larger timespans gives us more accurate estimates and thus more balanced scheduling.

The above algorithm is inspired by Linux's "Completely Fair Scheduler".

Note that the runtime estimate doesn't exactly correlate to "real" time:
- we increment the runtime when starting a fuzzer,
- the runtime doesn't count time that the fuzzer is running while the CFO is also compiling another task, and
- if a fuzzer completes in less than 1 second (the CFO's loop-sleep) then it still counts as 1 second of runtime.

# Follow-up

We should probably adjust the weights as well, e.g. not run the canary so often.

# Measurements

These are just 20 minute samples.

<details>
<summary>Cumulative estimated runtimes, pre-fair-scheduling</summary>

```
debug: commit=bb40482 fuzzer=canary                   runtime=11
debug: commit=bb40482 fuzzer=ewah                     runtime=5
debug: commit=bb40482 fuzzer=lsm_cache_map            runtime=206
debug: commit=bb40482 fuzzer=lsm_forest               runtime=302
debug: commit=bb40482 fuzzer=lsm_manifest_level       runtime=15
debug: commit=bb40482 fuzzer=lsm_manifest_log         runtime=15
debug: commit=bb40482 fuzzer=lsm_scan                 runtime=9
debug: commit=bb40482 fuzzer=lsm_segmented_array      runtime=13
debug: commit=bb40482 fuzzer=lsm_tree                 runtime=189
debug: commit=bb40482 fuzzer=storage                  runtime=13
debug: commit=bb40482 fuzzer=vopr_lite                runtime=27
debug: commit=bb40482 fuzzer=vopr_testing_lite        runtime=27
debug: commit=bb40482 fuzzer=vopr_testing             runtime=45
debug: commit=bb40482 fuzzer=vopr                     runtime=49
debug: commit=bb40482 fuzzer=vsr_free_set             runtime=12
debug: commit=bb40482 fuzzer=vsr_journal_format       runtime=5
debug: commit=bb40482 fuzzer=vsr_superblock_quorums   runtime=17
debug: commit=bb40482 fuzzer=vsr_superblock           runtime=13
debug: commit=3f4fe4f fuzzer=canary                   runtime=37
debug: commit=3f4fe4f fuzzer=vopr_lite                runtime=58
debug: commit=d7167fd fuzzer=canary                   runtime=39
debug: commit=d7167fd fuzzer=vopr_lite                runtime=88
debug: commit=d7167fd fuzzer=vopr                     runtime=278
```

</details>

<details>
<summary>Cumulative estimated runtimes, post-fair-scheduling</summary>

(Note that `lsm_tree` only ran once, just for a long time:)

```
debug: commit=bb40482 fuzzer=canary                   runtime=35
debug: commit=bb40482 fuzzer=ewah                     runtime=35
debug: commit=bb40482 fuzzer=lsm_cache_map            runtime=34
debug: commit=bb40482 fuzzer=lsm_forest               runtime=64
debug: commit=bb40482 fuzzer=lsm_manifest_level       runtime=35
debug: commit=bb40482 fuzzer=lsm_manifest_log         runtime=34
debug: commit=bb40482 fuzzer=lsm_scan                 runtime=41
debug: commit=bb40482 fuzzer=lsm_segmented_array      runtime=34
debug: commit=bb40482 fuzzer=lsm_tree                 runtime=104
debug: commit=bb40482 fuzzer=storage                  runtime=34
debug: commit=bb40482 fuzzer=vopr_lite                runtime=68
debug: commit=bb40482 fuzzer=vopr_testing_lite        runtime=69
debug: commit=bb40482 fuzzer=vopr_testing             runtime=84
debug: commit=bb40482 fuzzer=vopr                     runtime=68
debug: commit=bb40482 fuzzer=vsr_free_set             runtime=34
debug: commit=bb40482 fuzzer=vsr_journal_format       runtime=35
debug: commit=bb40482 fuzzer=vsr_superblock_quorums   runtime=35
debug: commit=bb40482 fuzzer=vsr_superblock           runtime=35
debug: commit=3f4fe4f fuzzer=canary                   runtime=121
debug: commit=3f4fe4f fuzzer=vopr_lite                runtime=239
debug: commit=d7167fd fuzzer=canary                   runtime=119
debug: commit=d7167fd fuzzer=vopr_lite                runtime=238
debug: commit=d7167fd fuzzer=vopr                     runtime=238
```

</details>

matklad commented on src/scripts/cfo.zig:
```diff
@@ -211,34 +211,41 @@ fn run_fuzzers(
+        const tasks = tasks: {
+            if (iteration_pull) {
+                // TODO: This is a race -- run_fuzzers_prepare_tasks() modifies the working
+                // directory, which may still be in use by a running fuzzer.
```
Oops, I forgot to remove this TODO in https://github.com/tigerbeetle/tigerbeetle/pull/2680, let's remove it here!

matklad commented on src/scripts/cfo.zig:
```diff
@@ -336,12 +357,26 @@ fn run_fuzzers(
     weight: []u32,
+    /// Estimated cumulative runtime of the corresponding task.
+    runtimes: []u64,
 };
```
It feels a bit weird that we do both an AoS and a Map here now. Perhaps its time to raise the level of abstraction here and do something like

```zig
const Tasks = struct {
    const Task = struct {
       ...
    };

    pub fn sample(tasks: *const Tasks) *Task {}
    pub fn get(tasks: *const Tasks, fuzzer: Fuzzer, commit: [40]u8) ?*Task {}
    pub fn get_or_insert(tasks: *const Tasks, task: *Task) *Task {}
    pub fn soft_remove_all() void {}
};
```

Though, I think the current code is fine, even if it was written with the idea that tasks are immutable.

matklad commented on src/scripts/cfo.zig:
```diff
@@ -467,11 +502,79 @@ fn run_fuzzers_prepare_tasks(shell: *Shell, gh_token: ?[]const u8) !Tasks {
+}
+
+/// Initialize the runtime of each task in `tasks`.
+fn run_fuzzers_prepare_tasks_runtime(tasks: Tasks, tasks_cache: Tasks) void {
```
I am starting to hate my convention of prefixing everything with `run_fuzzers` 🤣

matklad commented on src/scripts/cfo.zig:
```diff
@@ -267,6 +274,20 @@ fn run_fuzzers(
+                })) |task_index| {
+                    tasks.runtimes[task_index] += 1;
+                } else {
+                    // This is a leftover fuzzer from a now-cancelled task.
```
An alternative would be to make Tasks append-only, and mark inactive tasks (or, rather, track the generation number, and only consider tasks from the latest generation to be active)

matklad commented on src/scripts/cfo.zig:
```diff
@@ -467,11 +502,79 @@ fn run_fuzzers_prepare_tasks(shell: *Shell, gh_token: ?[]const u8) !Tasks {
+}
+
+/// Returns the index of the task with the minimal `runtime / weight`.
+fn run_fuzzers_tasks_sample(tasks: Tasks) usize {
```
Feels like a clear-cut method of Tasks struct?

matklad reviewed (commented):
LGTM!

Consider if you might want to do a refactoring pass over the entire thing, as I feel that, if we write the code from scratch, we would do some of the details differently. But looks good as is!

sentientwaffle committed:
CFO: Refactor tasks, and fix another schedule weight bug

## Bug

Consider if you have a couple PR's being fuzzed (along with main), and then one of them gets merged.

When both PR's were being fuzzed, we split our fuzz time 50:50 between main and PR's.
So each PR got 25% of runtime. This is currently accomplished with task weight adjustments.

But when one of the PR's is merged, the fuzz time is still split 50:50 -- the remaining PR now has much higher (relative) weight than before.
That increase in weight means that the denominator of the `runtime / weight` is higher, so the remaining PR's virtual runtime is suddenly cut in half.
This would cause it to starve main until it catches up.

## Fix

Accumulate virtual runtime (`runtime_virtual`) as `+= 1000 / weight`, so that changes to the weight don't cause massive virtual runtime imbalances.

We still track regular `runtime` just because that is more meaningful for logging. But `runtime` isn't used for scheduling anymore.

sentientwaffle commented:
Refactored as you suggest!

---

Also while refactoring I found a bug in the scheduling, which is now fixed:

Consider if you have a couple PR's being fuzzed (along with main), and then one of them gets merged.

When both PR's were being fuzzed, we split our fuzz time 50:50 between main and PR's.
So each PR got 25% of runtime. This is currently accomplished with task weight adjustments.

But when one of the PR's is merged, the fuzz time is still split 50:50 -- the remaining PR now has much higher (relative) weight than before. That increase in weight means that the denominator of the `runtime / weight` is higher, so the remaining PR's virtual runtime is suddenly cut in half.

This would cause the PR to starve main until it catches up.

Fix:

Accumulate virtual runtime (`runtime_virtual`) as `+= 1000 / weight`, so that changes to the weight don't cause massive virtual runtime imbalances.

We still track regular `runtime` just because that is more meaningful for logging. But `runtime` isn't used for scheduling anymore.
